### ISSUE TITLE
Define and implement new interactor syntax

### ISSUE THREAD

**Issue by cowboyd** [bigtestjs/server#35]
Part of #28 

After two years of learning, it's time to re-think the interactor API to be more intuitive based on users' existing knowledge of the DOM and the principles of component oriented design.

To do this we'll model it as a queryable, continuous graph of strongly typed relationships rather than an object instance created out of thin air (every interactor is ultimately traceable to the root) Ultimately this will be more understandable and ultimately more composable.

Get all the button on the page:

```js

import { Button } from `@bigtest/interactor`;

let buttons = Button.all(); //=> Iterable<Button>

let first = buttons.first();
```

```js
class Form extends Interactor {
  cancel = Button.first();
  submit = Button.last();

  input = TextField.first('[data-test-input]');

  errors = Validation.
      .where('[data-validation-error]')
      .where(':visible');
}
```

In this way, it takes on the characteristics of a `Link` relationship or an `ActiveRecord` relationship (for those familiar with those technologies).

These query methods do  not actually look up the button when the method is called, it only defines the relationship of the button to the structure of the document or sub-document. For example, the only time that the DOM is actually queried for existence of the button would be at the time an action method like`click()` is invoked. Even in that situtation however, the actions will be convergent which means that they will perform themselves whenever the button becomes available, and not fail if they aren't at the exact moment that the action is invoked)

This is only a straw man proposal, and definition of the Interactor syntax is part of the scope of this issue

**Comment by JohnC-80** [comment]
When it comes to finding elements and interacting with applications as a user does, it seems like the fairly recent "DOM Testing Library" has some interesting [methods/queries](https://testing-library.com/docs/dom-testing-library/api-queries#queries)... curious if you guys have any thoughts about these - would similar functionality (findByLabelText, etc) be a good addition to Interactor?  General inclusion of a11y attributes seems like a nice detail. Possibly something that could be used in place of a selector for actions `clickable(text('Submit'))`.

**Comment by JohnC-80** [comment]
When I was starting with BigTest, the `-ables` were a stumbling point... would a `will-` prefix for action creators work better? `willClick` `willFocus` `willBlur` ?

**Comment by cowboyd** [comment]
Agree, that is a good starting point, and I think we should absolutely be able to add queries like that.

Also, I think we should get rid of the "ables" entirely. I think we missed the proper unit of composition. It isn't the properties of the interactors, but the interactors themselves that are the composable unit, just like components.

For example, I'm imagining everything will be subclassed from an `Element` interactor. And really, you should be able to use Element interactor for almost anything:

```js
const input = Element.first('input[data-test-username]');
const submit = Element.first('button');

await input.focus();
await input.type('cowboyd');
await submit.click();
```

If all you wanted to do was use `Element` interactors all day, you would be fine, but if you want to have special types, you would compose them just like you would normal objects.

```js
class LookupUserForm extends Element {
  static scope = 'form[data-test-user-form]';

  username = Element.first('input[data-test-username]');
  submit = Element.first('button');

  async enterAndSubmit(username) {
    await this.input.focus();
    await this.input.type('cowboyd');
    await this.submit.click();
  }
}
```

**Comment by JohnC-80** [comment]
From the appearance, it looks like there's going to be a great deal of `.first` used with this API...even when I just expect a single element! I'm curious about the problem that gets solved with the iterable return value... it almost seems like a `collection`-ish API but used for single elements. Seeing your initial example with the form, where `Button.first()` is cancel and `Button.last()` is submit makes my eye twitch, as it would have to be updated at the rate at which the UX designer changes their mind on order of these types of actions. Does the iterable structure adequately decouple from DOM structure?

**Comment by JohnC-80** [comment]
What do you think of having both `Button` and `Buttons` - where `Button` could resolve to `Buttons.first()` and `Buttons` returns the iterable?

**Comment by pittst3r** [comment]
I've made this issue into a quest issue for better collaboration. Please see the **Action items** list above for the sub-issues. We should try to keep discussions relevant to those issues in their respective issues.

It may be good to move API-design related discussions to [Draft new API using type declarations](https://github.com/bigtestjs/interactor/issues/75). Or we can keep it here since it's already started.
